lib/server/controllers/file_controller.rb

```diff
@@ -72,36 +74,116 @@ def rename
+    raise Etna::BadRequest, 'At least one revision required' unless revisions.length > 0
+    raise Etna::BadRequest, 'All revisions require valid "source" parameter' unless revisions.all? { |rev| rev.source.valid?}
+    raise Etna::BadRequest, 'All revisions require valid "dest" parameter' unless revisions.all? { |rev| rev.dest.valid?}
 
-    new_bucket = require_bucket(@params[:new_bucket_name])
+    # In bulk copy mode, we validate the buckets in bulk
+    #   since they aren't part of the path (are part of revisions),
+    #   and trying to minimize database hits by rejecting early, if
+    #   possible.
+    all_source_bucket_names = revisions.map {|rev| rev.source.bucket_name}.uniq
+    all_dest_bucket_names = revisions.map {|rev| rev.dest.bucket_name}.uniq
 
-    new_folder = require_folder(new_bucket, new_folder_path)
+    hmac = @request.env['etna.hmac']
 
-    raise Etna::Forbidden, 'Folder is read-only' if new_folder&.read_only?
+    user_authorized_bucket_names = Metis::Bucket.where(
+      project_name: @params[:project_name],
+      owner: ['metis', hmac.id.to_s],
+      name: all_source_bucket_names + all_dest_bucket_names
+    ).all.select{|b| b.allowed?(@user, hmac)}.
+      map {|bucket| bucket.name}
+
+    raise Etna::Forbidden, 'User does not have access to all source buckets' unless revisions.all? {
+      |rev| rev.valid?('source_bucket_access', user_authorized_bucket_names)}
+    raise Etna::Forbidden, 'User does not have access to all dest buckets' unless revisions.all? {
+      |rev| rev.valid?('dest_bucket_access', user_authorized_bucket_names)}
+
+    revisions.each do |revision|
+      validate_copy_revision(revision)
+    end
+
+    # If we've gotten here, every revision looks good and we can execute them!
+    new_files = []
+    revisions.each do |revision|
+
+      new_file = Metis::File.copy({
+        project_name: @params[:project_name],
+        source_file: revision.source.file,
+        dest_file_path: revision.dest.file_path,
+        dest_bucket_name: revision.dest.bucket_name,
+        user: @user
+      })
+
+      new_files.push(new_file.to_hash(@request))
+    end
+    success_json(files: new_files)
+  end
 
-    raise Etna::Forbidden, 'Cannot copy over existing file' if Metis::File.exists?(new_file_name, new_bucket, new_folder)
+  private
 
-    raise Etna::Forbidden, 'Cannot copy over existing folder' if  Metis::Folder.exists?(new_file_name, new_bucket, new_folder)
+  def validate_copy_revision(revision)
+    file = revision.source.file
 
-    new_file = Metis::File.create(
-      project_name: @params[:project_name],
-      file_name: new_file_name,
-      folder_id: new_folder&.id,
-      bucket: new_bucket,
-      author: Metis::File.author(@user),
-      data_block: file.data_block
-    )
-    success_json(files: [ new_file.to_hash(@request) ])
+    raise Etna::Error.new("File #{revision.source.path} not found", 404) unless file&.has_data?
```

Comment:
This method definitely seems like it should be on the revision object. Generally I think it seems preferable to have all of the raising occurring in the controller, and have service objects handle the problem of determining error states. In this case, I think it's especially important not to raise until all of the errors have been collected - since the user may have submitted perhaps hundreds or thousands of revisions, they shouldn't have to discover and repair them one-by-one. The revision object can just collect the errors in an @errors method, which the controller can retrieve and compose an error message containing them. This probably allows the controller to only raise in one place, although it will almost certainly be forced to label everything as "bad request" rather than returning individual statuses.

Reply:
K, that makes sense. I'll see if I can work that out.